From 78ec5b4df54ffb5895ae282f7746cd56b91276dd Mon Sep 17 00:00:00 2001
From: Girija P <gpandiya@qti.qualcomm.com>
Date: Mon, 2 Feb 2026 14:34:06 +0530
Subject: [PATCH] Onboarding bug fixes for 2 3link MLO only case Reserve
 mld_unit 0 for bsta only Don't configure wds_sta from mesh

Signed-off-by: Girija P <gpandiya@qti.qualcomm.com>
---
 agent/src/beerocks/slave/agent_db.cpp         | 10 ++++++----
 .../slave/tasks/ap_autoconfiguration_task.cpp |  4 ++++
 .../beerocks/bwl/include/bwl/nl80211_client.h |  8 +++++++-
 .../bwl/include/bwl/nl80211_client_impl.h     |  2 +-
 .../bwl/nl80211/ap_wlan_hal_nl80211.cpp       | 19 ++++++++++++++-----
 .../bwl/shared/nl80211_client_impl.cpp        | 10 ++++++----
 .../beerocks/bwl/whm/nl80211_client_whm.cpp   |  4 ++--
 common/beerocks/bwl/whm/nl80211_client_whm.h  |  2 +-
 8 files changed, 41 insertions(+), 18 deletions(-)

diff --git a/agent/src/beerocks/slave/agent_db.cpp b/agent/src/beerocks/slave/agent_db.cpp
index 72ed667cf..7b7ce6275 100644
--- a/agent/src/beerocks/slave/agent_db.cpp
+++ b/agent/src/beerocks/slave/agent_db.cpp
@@ -35,6 +35,8 @@ AgentDB::sRadio *AgentDB::add_radio(const std::string &front_iface_name,
         return nullptr;
     }
 
+    std::string local_back_iface_name = back_iface_name;
+
     if (!front_iface_name.empty() && radio(front_iface_name)) {
         LOG(DEBUG) << "Radio entry of front iface " << front_iface_name
                    << " already exists. Not adding.";
@@ -42,12 +44,12 @@ AgentDB::sRadio *AgentDB::add_radio(const std::string &front_iface_name,
     }
 
     if (!back_iface_name.empty() && radio(back_iface_name)) {
-        LOG(DEBUG) << "Radio entry of back iface " << back_iface_name
-                   << " already exists. Not adding.";
-        return nullptr;
+        LOG(ERROR) << "Radio entry of back iface " << back_iface_name
+                   << " already exists. Not adding this back iface again to another radio.";
+        local_back_iface_name = "";
     }
 
-    m_radios.emplace_back(front_iface_name, back_iface_name);
+    m_radios.emplace_back(front_iface_name, local_back_iface_name);
     m_radios_list.push_back(&m_radios.back());
 
     return &m_radios.back();
diff --git a/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp b/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp
index 1ff7bb2c6..7a75ca01a 100644
--- a/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp
+++ b/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp
@@ -3411,6 +3411,10 @@ int8_t ApAutoConfigurationTask::find_available_ap_mld_unit()
     auto db = AgentDB::get();
 
     std::unordered_set<int8_t> used_mld_units;
+    
+    // Reserve unit 0 for bSTA MLD (since find_available_bsta_mld_unit always returns 0)
+    used_mld_units.insert(0);
+    
     for (auto ap_mld_conf : db->ap_mld_configurations) {
         used_mld_units.insert(ap_mld_conf.mld_config.mld_unit);
     }
diff --git a/common/beerocks/bwl/include/bwl/nl80211_client.h b/common/beerocks/bwl/include/bwl/nl80211_client.h
index acaa1f8d3..ff0d7f923 100644
--- a/common/beerocks/bwl/include/bwl/nl80211_client.h
+++ b/common/beerocks/bwl/include/bwl/nl80211_client.h
@@ -46,6 +46,12 @@ static constexpr size_t he_mcs_set_size = beerocks::message::HE_MCS_SET_NL802_SI
 
 namespace bwl {
 
+// Return value macros for create_new_interface API
+#define NEW_IFACE_SUCCESS 0
+#define NEW_IFACE_EXISTS 1
+#define NEW_IFACE_FAIL -1
+
+
 /**
  * @brief NL80211 client interface.
  *
@@ -616,7 +622,7 @@ public:
     virtual bool send_delba(const std::string &interface_name, const sMacAddr &dst,
                             const sMacAddr &src, const sMacAddr &bssid) = 0;
 
-    virtual bool create_new_interface(std::string &vap_name ) =0;
+    virtual int create_new_interface(std::string &vap_name ) =0;
 
     virtual bool get_bss_link_id(std::string &interface_name,std::string &link_mac,int8_t &link_id) = 0;
 
diff --git a/common/beerocks/bwl/include/bwl/nl80211_client_impl.h b/common/beerocks/bwl/include/bwl/nl80211_client_impl.h
index d4ed61439..b464dadb2 100644
--- a/common/beerocks/bwl/include/bwl/nl80211_client_impl.h
+++ b/common/beerocks/bwl/include/bwl/nl80211_client_impl.h
@@ -181,7 +181,7 @@ public:
     bool send_delba(const std::string &interface_name, const sMacAddr &dst, const sMacAddr &src,
                     const sMacAddr &bssid) override;
 
-    virtual bool create_new_interface(std::string &vap_name) override;
+    virtual int create_new_interface(std::string &vap_name) override;
 
     virtual bool get_bss_link_id(std::string &interface_name,std::string &link_mac,int8_t &link_id) override;
 
diff --git a/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp b/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp
index 82c54bfca..e6300d75b 100644
--- a/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp
+++ b/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp
@@ -1541,9 +1541,15 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
             bss_it.old_iface_name = new_vap_name;
 
             if ((mld_id_val < 0) || ((mld_id_val >= 0) && (bss_name_changed == false))) {
-                if (!m_nl80211_client->create_new_interface(new_vap_name)) {
-                    LOG(ERROR) << "Unable to create new interface ";
-                    continue;
+                int ret_val =  m_nl80211_client->create_new_interface(new_vap_name);
+                if (ret_val != NEW_IFACE_SUCCESS) {
+                    if (ret_val == NEW_IFACE_EXISTS) {
+                        LOG(INFO) << "There is existing interface," << new_vap_name << " so need to add link instead of creating new BSS";
+                        bss_name_changed = true;
+                    } else {
+                        LOG(ERROR) << "Unable to create new interface ";
+                        continue;
+                    }
                 }
             }
 
@@ -1724,7 +1730,8 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
                 // finally enable the vap (remove any previously set start_disabled)
                 conf.set_create_vap_value(vap_id, "start_disabled", "");
 
-                conf.set_create_vap_value(vap_id, "wds_sta", "1");
+//  multi_ap and wds_sta cannot co-exist for AP 
+//                conf.set_create_vap_value(vap_id, "wds_sta", "1");
 
                 if (mld_id_val >= 0) {
                     std::string new_bss_name = get_ifname_wrt_mld_grp(mld_id_val, bss,bss_it.ssid);
@@ -3115,7 +3122,9 @@ bool ap_wlan_hal_nl80211::create_new_hostapd_conf_file(std::string &ifname, son:
         LOG(ERROR) << "Failed to set the authentication/encryption parameters!";
         return false;
     }
-    conf.set_create_vap_value(vap_id, "wds_sta", "1");
+//  multi_ap and wds_sta cannot co-exist for AP    
+//    conf.set_create_vap_value(vap_id, "wds_sta", "1");
+
     if(!m_radio_info.wps_pushbutton_configured && bss_conf.fronthaul){
         conf.set_create_vap_value(vap_id, "wps_state", bss_conf.fronthaul ? "2" : "");
         conf.set_create_vap_value(vap_id, "wps_independent", bss_conf.fronthaul ? "1" : "");
diff --git a/common/beerocks/bwl/shared/nl80211_client_impl.cpp b/common/beerocks/bwl/shared/nl80211_client_impl.cpp
index 4fbdb4674..f7fab946c 100644
--- a/common/beerocks/bwl/shared/nl80211_client_impl.cpp
+++ b/common/beerocks/bwl/shared/nl80211_client_impl.cpp
@@ -1714,22 +1714,22 @@ bool nl80211_client_impl::send_delba(const std::string &interface_name, const sM
                                             [&](struct nl_msg *msg) {});
 }
 
-bool nl80211_client_impl::create_new_interface(std::string &vap_name)
+int nl80211_client_impl::create_new_interface(std::string &vap_name)
 {
     LOG(ERROR) << "Create new interface: " << vap_name;
     if (!m_socket) {
         LOG(ERROR) << "Socket is NULL!";
-        return false;
+        return NEW_IFACE_FAIL;
     }
 
     // Get the interface index for given interface name
     int iface_index = if_nametoindex(vap_name.c_str());
     if (0 != iface_index) {
         LOG(DEBUG) << "Interface " << vap_name << " is already present with iface_index: " << iface_index ;
-        return true;
+        return NEW_IFACE_EXISTS;
     }
 
-    return m_socket.get()->send_receive_msg(
+    bool result = m_socket.get()->send_receive_msg(
         NL80211_CMD_NEW_INTERFACE, 0,
         [&](struct nl_msg *msg) -> bool {
             if(nla_put_string(msg, NL80211_ATTR_IFNAME,vap_name.c_str()) < 0){
@@ -1760,6 +1760,8 @@ bool nl80211_client_impl::create_new_interface(std::string &vap_name)
                 LOG(ERROR) << "Success parse netlink message!";
 
         });
+    
+    return result ? NEW_IFACE_SUCCESS : NEW_IFACE_FAIL;
 }
 
 bool nl80211_client_impl::remove_vap_interface(std::string &iface)
diff --git a/common/beerocks/bwl/whm/nl80211_client_whm.cpp b/common/beerocks/bwl/whm/nl80211_client_whm.cpp
index e755ff864..5b91d7b56 100644
--- a/common/beerocks/bwl/whm/nl80211_client_whm.cpp
+++ b/common/beerocks/bwl/whm/nl80211_client_whm.cpp
@@ -101,10 +101,10 @@ bool nl80211_client_whm::get_sta_info(const std::string &interface_name,
     return true;
 }
 
-bool nl80211_client_whm::create_new_interface(std::string &vap_name)
+int nl80211_client_whm::create_new_interface(std::string &vap_name)
 {
     LOG(TRACE) << __func__ << " - NOT IMPLEMENTED";
-    return false;
+    return NEW_IFACE_FAIL;
 }
 
 bool nl80211_client_whm::get_bss_link_id(std::string &interface_name, std::string &link_mac, int8_t &link_id)
diff --git a/common/beerocks/bwl/whm/nl80211_client_whm.h b/common/beerocks/bwl/whm/nl80211_client_whm.h
index f8926bba3..7c14c5b38 100644
--- a/common/beerocks/bwl/whm/nl80211_client_whm.h
+++ b/common/beerocks/bwl/whm/nl80211_client_whm.h
@@ -80,7 +80,7 @@ public:
                       sta_info &sta_info) override;
 
     // Stub implementations for pure virtual functions from base class
-    virtual bool create_new_interface(std::string &vap_name) override;
+    virtual int create_new_interface(std::string &vap_name) override;
     virtual bool get_bss_link_id(std::string &interface_name, std::string &link_mac, int8_t &link_id) override;
     virtual bool get_iface_with_link_idx(const std::string &iface_radio_idxd, std::string &iface_link_idxd) override;
     virtual bool getMLDLinkIfaceMacAddr(std::string &interface_name, std::string &mac_out) override;
-- 
2.34.1

